### ISSUE TITLE
4DS: NaNs in inputs

### ISSUE THREAD

**Issue by snjsnj** [UK-Digital-Heart-Project/4Dsurvival#5]
@lisurui6 

**Brief Description:**

While training the 4DS network ([demo_validateDL.py](https://github.com/UK-Digital-Heart-Project/4Dsurvival/blob/master/demo/demo_validateDL.py)) using PH data from the cardiac drive, _'NaNs detected in inputs, please correct or drop'_ error is observed:

![image](https://user-images.githubusercontent.com/16711478/110918916-9ad1f380-8313-11eb-9242-c51f91b3e36a.png)

**About the screenshot:**
1. I have suppressed the N/W training progress information.
2. The training starts fine with step (1)
3. For the first bootsrap sample, the error is observed in step (2a), hyperparameter search.
[concordance.py](https://github.com/CamDavidsonPilon/lifelines/blob/master/lifelines/utils/concordance.py)
4. The error is 'NaNs detected in inputs, please correct or drop'

**My understanding:**

While finding the hyperparameters, the inputs are becoming NaNs.
Possible causes:
1. Exploding/vanishing gradient problem
2. Seeding the solver properly
3. Need of changing the inputs passed to the _DL_single_run()_ inside [hypersearch_DL()](https://github.com/UK-Digital-Heart-Project/4Dsurvival/blob/master/survival4D/hypersearch.py)

**Reproducing the error:**

1. >screen
2. Run the latest docker image with appropriate mounting of the volume for the following file: _$PWD_/cardiac/for_ghalib/backup/home/gbello@isd.csc.mrc.ac.uk/gbello/RWorkspace1/mesh_motiondescriptor_PHAB0267Adecmesh_CTEPH302.pkl: 
>nvidia-docker run -ti --rm -v  _$PWD_/cardiac/for_ghalib/backup/home/gbello@isd.csc.mrc.ac.uk/gbello/RWorkspace1:/data lisurui6/4dsurvival-gpu:latest
3. >cd demo
4. >CUDA_VISIBLE_DEVICES=_'available_device'_ python3 demo_validateDL.py
5. Wait.

**Comment by lisurui6** [comment]
How long do I have to wait to see NaN error occur?

**Comment by lisurui6** [comment]
I would need more information. Does the loss in step 2a become NaN at some point? What iteration?

**Comment by lisurui6** [comment]
A few points to note down:

1. I don't think such extensive search of hyperparameters is useful in practice, and without it would surely make it more efficient to execute the code. I think once they are found in step 1a, the hyperparameters can be used in the bootstrap sampling steps. Things like drop out rate, number of hidden units, is not that sensitive. Learning rate isn't sensitive as well, if Adam optimiser is used. So there is little point in repeating the search. 
2.  Most likely the NaN is from exploding gradients. Adding batch norm layers definitively is recommended, and will solve this issue. Notice to add them after linear layer before nonlinear activation (and of course not in the output layer). 
3. The repo uses keras==2.2.2 and tf1. Highly recommended to upgrade to tf2 at least. Most recommended to switch to pytorch. It will be much easier to debug.

**Comment by doregan** [comment]
Lifelines function had a bugfix in April 2019 as it had been silently converting NaNs in the event vector to True. In the current version these now produce an error - which we are picking up when we ran the code. Reducing dropout search range to [0.1,0.5] allows code to run with lifelines v 0.23.9 and achieves similar discrimination performance.  
![image (8)](https://user-images.githubusercontent.com/5120555/111474838-60f25a00-8724-11eb-90d8-333582de264d.png)
